aspects/schema.go

```diff
@@ -0,0 +1,328 @@
+	switch typ {
+	case "map":
+		return &mapSchema{topSchema: s}, nil
+	case "string":
+		return &stringSchema{}, nil
+	default:
+		return nil, fmt.Errorf("cannot parse type %q: unknown", typ)
+	}
+}
+
+type mapSchema struct {
+	// topSchema is the schema for the top-level schema which contains the user types.
+	topSchema *StorageSchema
+
+	// entrySchemas map keys that can the map can contain to their expected types.
+	// Alternatively, the schema can instead key and/or value types.
+	entrySchemas map[string]Schema
+
+	// valueSchema validates that the map's values match a certain type.
+	valueSchema Schema
+
+	// keySchema validates that the map's key match a certain type.
+	keySchema Schema
+
+	// requiredCombs holds combinations of keys that an instance of the map is
+	// allowed to have.
+	requiredCombs [][]string
+}
+
+func (v *mapSchema) Validate(raw []byte) error {
+	var mapValue map[string]json.RawMessage
+	if err := json.Unmarshal(raw, &mapValue); err != nil {
+		return err
+	}
+
+	var missing bool
+	for _, required := range v.requiredCombs {
+		missing = false
+		for _, key := range required {
+			if _, ok := mapValue[key]; !ok {
+				missing = true
+				break
+			}
+		}
+
+		if !missing {
+			// matched possible combination of required keys so we can stop
+			break
+		}
+	}
+
+	if missing {
+		return fmt.Errorf(`cannot find required combinations of keys`)
+	}
+
+	if v.entrySchemas != nil {
+		for key, val := range mapValue {
+			if validator, ok := v.entrySchemas[key]; ok {
+				if err := validator.Validate(val); err != nil {
+					return err
+				}
+			}
+		}
+
+		// all required entries are present and validated
+		return nil
+	}
+
+	if v.keySchema != nil {
+		for k := range mapValue {
+			rawKey, err := json.Marshal(k)
+			if err != nil {
+				return err
+			}
+
+			if err := v.keySchema.Validate(rawKey); err != nil {
+				return err
+			}
+
+		}
+	}
+
+	if v.valueSchema != nil {
+		for _, val := range mapValue {
+			if err := v.valueSchema.Validate(val); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (v *mapSchema) parseConstraints(raw json.RawMessage) error {
+	var constraints map[string]json.RawMessage
+	err := json.Unmarshal(raw, &constraints)
+	if err != nil {
+		return err
+	}
+
+	rawRequired, ok := constraints["required"]
+	if ok {
+		var requiredCombs [][]string
+		if err := json.Unmarshal(rawRequired, &requiredCombs); err != nil {
+			var typeErr *json.UnmarshalTypeError
+			if !errors.As(err, &typeErr) {
+				return fmt.Errorf(`cannot unmarshal map's "required" constraint: %v`, err)
+			}
+
+			var required []string
+			if err := json.Unmarshal(rawRequired, &required); err != nil {
+				return fmt.Errorf(`cannot unmarshal map's "required" constraint: %v`, err)
+			}
+
+			v.requiredCombs = [][]string{required}
+		} else {
+			v.requiredCombs = requiredCombs
+		}
+	}
+
+	// maps can have a "schema" constraint with types for specific entries
+	if rawEntries, ok := constraints["schema"]; ok {
```

Comment:
I think we want to support only this combinations of constaints, either:

* schema and optionally required
* or keys and/or values

so if "schema" appears "keys" cannot be used etc